File: frontend/src/api/sync.ts
/** 同步管理模块 REST 客户端（对齐后端 api接口清单 §六 / schemas/sync.py）。 */
import { api } from "./client";

// ---- GET /sync/status ----

export interface SyncStats {
  total_chunks: number;
  code_chunks: number;
  doc_chunks: number;
  stale_docs: number;
  total_relations: number;
  total_anchors: number;
}

export interface SyncStatusResponse {
  status: string;
  last_sync_at: string | null;
  last_commit: string | null;
  stats: SyncStats;
}

// ---- GET /sync/tasks ----

export interface ChangeSummary {
  added: number;
  modified: number;
  deleted: number;
}

export interface RollbackDetail {
  chunks_rolled_back: number;
  chunks_restored: number;
  relations_restored: number;
  anchors_restored: number;
  stale_anchors_cleared: number;
}

export interface SyncTaskItem {
  task_id: number;
  type: string;
  commit: string;
  status: string;
  changes: ChangeSummary;
  started_at: string | null;
  finished_at: string | null;
  duration_ms: number | null;
  doc_pr_url: string | null;
  doc_pr_status: string | null;
  source_commit: string | null;
  rollback_detail: RollbackDetail | null;
  triggered_by: string | null;
}

export interface SyncTaskListResponse {
  total: number;
  items: SyncTaskItem[];
}

export interface ChangeDetailItem {
  chunk_id: string;
  file: string | null;
  change_type: string;
  rollback_source_commit: string | null;
}

export interface SyncTaskDetailResponse extends SyncTaskItem {
  change_details: ChangeDetailItem[];
  errors: Array<{ file?: string; change?: string; error: string }>;
}

// ---- GET /sync/rollbacks ----

export interface RollbackItem {
  rollback_id: number;
  rollback_commit: string;
  source_commit: string;
  chunks_rolled_back: number;
  chunks_restored: number;
  chunks_deleted: number;
  relations_restored: number;
  anchors_restored: number;
  stale_anchors_cleared: number;
  doc_pr_closed: string | null;
  triggered_by: string;
  status: string;
  created_at: string;
}

export interface RollbackListResponse {
  total: number;
  items: RollbackItem[];
}

// ---- POST /sync/trigger ----

export type SyncType = "FULL" | "INCREMENTAL";

export interface TriggerRequest {
  type: SyncType;
  target_commit?: string | null;
}

export interface TriggerResponse {
  task_id: number;
  status: string;
  message: string;
}

// ---- GET /sync/change-history/{chunk_id} ----

export interface ChangeHistoryItem {
  commit: string;
  change_type: string;
  date: string | null;
  old_hash: string | null;
  new_hash: string | null;
  is_rollback_related: boolean;
}

export interface ChangeHistoryResponse {
  chunk_id: string;
  file: string | null;
  history: ChangeHistoryItem[];
}

// ---- 调用 ----

export const getSyncStatus = () =>
  api.get<SyncStatusResponse>("/v1/sync/status").then((r) => r.data);

export const listSyncTasks = (params: { type?: string; status?: string; page?: number; page_size?: number } = {}) =>
  api.get<SyncTaskListResponse>("/v1/sync/tasks", { params }).then((r) => r.data);

export const getSyncTask = (taskId: number) =>
  api.get<SyncTaskDetailResponse>(`/v1/sync/tasks/${taskId}`).then((r) => r.data);

export const listRollbacks = (params: { page?: number; page_size?: number } = {}) =>
  api.get<RollbackListResponse>("/v1/sync/rollbacks", { params }).then((r) => r.data);

export const triggerSync = (body: TriggerRequest) =>
  api.post<TriggerResponse>("/v1/sync/trigger", body).then((r) => r.data);

export const getChangeHistory = (chunkId: string) =>
  api.get<ChangeHistoryResponse>(`/v1/sync/change-history/${chunkId}`).then((r) => r.data);

File: backend/app/schemas/sync.py
"""同步管理模块的请求/响应 schema（对齐 api接口清单 §六）。"""
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel

# ---- GET /status ----

class SyncStats(BaseModel):
    total_chunks: int
    code_chunks: int
    doc_chunks: int
    stale_docs: int
    total_relations: int
    total_anchors: int


class SyncStatusResponse(BaseModel):
    status: str
    last_sync_at: datetime | None = None
    last_commit: str | None = None
    stats: SyncStats


# ---- GET /tasks ----

class ChangeSummary(BaseModel):
    added: int
    modified: int
    deleted: int


class RollbackDetail(BaseModel):
    chunks_rolled_back: int
    chunks_restored: int
    relations_restored: int
    anchors_restored: int
    stale_anchors_cleared: int


class SyncTaskItem(BaseModel):
    task_id: int
    type: str
    commit: str
    status: str
    changes: ChangeSummary
    started_at: datetime | None = None
    finished_at: datetime | None = None
    duration_ms: int | None = None
    doc_pr_url: str | None = None
    doc_pr_status: str | None = None
    source_commit: str | None = None
    rollback_detail: RollbackDetail | None = None
    triggered_by: str | None = None


class SyncTaskListResponse(BaseModel):
    total: int
    items: list[SyncTaskItem]


class ChangeDetailItem(BaseModel):
    chunk_id: str
    file: str | None = None
    change_type: str
    rollback_source_commit: str | None = None


class SyncTaskDetailResponse(SyncTaskItem):
    change_details: list[ChangeDetailItem] = []
    errors: list[dict] = []


# ---- GET /rollbacks ----

class RollbackItem(BaseModel):
    rollback_id: int
    rollback_commit: str
    source_commit: str
    chunks_rolled_back: int
    chunks_restored: int
    chunks_deleted: int
    relations_restored: int
    anchors_restored: int
    stale_anchors_cleared: int
    doc_pr_closed: str | None = None
    triggered_by: str
    status: str
    created_at: datetime


class RollbackListResponse(BaseModel):
    total: int
    items: list[RollbackItem]


# ---- POST /trigger ----

class TriggerRequest(BaseModel):
    type: Literal["FULL", "INCREMENTAL"]
    target_commit: str | None = None


class TriggerResponse(BaseModel):
    task_id: int
    status: str
    message: str


# ---- GET /change-history/{chunk_id} ----

class ChangeHistoryItem(BaseModel):
    commit: str
    change_type: str
    date: datetime | None = None
    old_hash: str | None = None
    new_hash: str | None = None
    is_rollback_related: bool = False


class ChangeHistoryResponse(BaseModel):
    chunk_id: str
    file: str | None = None
    history: list[ChangeHistoryItem]
